File: csharp/src/WordFrequencyCounter/WordCounter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace WordFrequencyCounter;

internal static class WordCounter
{
    private const int DefaultMaxWord = 64;
    private const int EstimatedBytesPerUniqueWord = 32;
    private const int MaxWordLimit = 1024;
    private const int MinWord = 4;

    internal static Result CountBytes(ReadOnlySpan<byte> bytes, int top, int maxWord)
    {
        Accumulator accumulator = new(NormalizeMaxWord(maxWord), EstimatedUniqueWords(bytes));
        accumulator.AddBytes(bytes);
        return accumulator.Finish(top);
    }

    private static bool IsLetter(byte value) => value is (>= (byte) 'A' and <= (byte) 'Z') or (>= (byte) 'a' and <= (byte) 'z');

    private static byte LowerAscii(byte value) => value is >= (byte) 'A' and <= (byte) 'Z' ? (byte) (value + 32) : value;

    private static int EstimatedUniqueWords(ReadOnlySpan<byte> bytes) => bytes.Length / EstimatedBytesPerUniqueWord;

    private static int NormalizeMaxWord(int value) =>
        value switch
        {
            0 => DefaultMaxWord,
            < MinWord => MinWord,
            > MaxWordLimit => MaxWordLimit,
            _ => value,
        };

    private sealed class Accumulator
    {
        private readonly Dictionary<string, ulong> _counts;
        private readonly int _maxWord;
        private readonly StringBuilder _word;
        private ulong _total;

        internal Accumulator(int maxWord, int estimatedUniqueWords)
        {
            _counts = new Dictionary<string, ulong>(estimatedUniqueWords, StringComparer.Ordinal);
            _maxWord = maxWord;
            _word = new StringBuilder(capacity: Math.Min(maxWord, DefaultMaxWord));
        }

        internal void AddBytes(ReadOnlySpan<byte> bytes)
        {
            foreach (byte value in bytes)
            {
                if (IsLetter(value))
                {
                    if (_word.Length < _maxWord)
                    {
                        _word.Append((char) LowerAscii(value));
                    }
                    continue;
                }

                if (_word.Length > 0)
                {
                    AddWord();
                }
            }
        }

        internal Result Finish(int top)
        {
            if (_word.Length > 0)
            {
                AddWord();
            }

            List<Entry> entries = [.. _counts.Select(pair => new Entry(pair.Key, pair.Value))];
            entries.Sort(CompareEntries);
            if (entries.Count > top)
            {
                entries.RemoveRange(top, entries.Count - top);
            }

            return new Result(_total, _counts.Count, entries);
        }

        private static int CompareEntries(Entry left, Entry right)
        {
            int countOrder = right.Count.CompareTo(left.Count);
            return countOrder != 0 ? countOrder : string.CompareOrdinal(left.Word, right.Word);
        }

        private void AddWord()
        {
            string key = _word.ToString();
            _counts[key] = _counts.TryGetValue(key, out ulong count) ? count + 1 : 1;
            _total++;
            _word.Clear();
        }
    }
}

File: php/src/WordCounter.php
<?php

declare(strict_types=1);

namespace WordFrequencyCounter;

final class WordCounter
{
    private const int ORACLE_DEFAULT_MAX_WORD = 64;
    private const int MAX_WORD_LIMIT = 1024;
    private const int MIN_WORD = 4;

    public static function countBytes(string $bytes, int $top, int $maxWord): Result
    {
        $maxWord = self::normalizeMaxWord($maxWord);

        /** @phpstan-var array<string, int> $counts */
        $counts = [];
        $word = '';
        $total = 0;
        $length = strlen($bytes);

        for ($index = 0; $index < $length; $index++) {
            $lower = ord($bytes[$index]) | 32;

            if ($lower >= 97 && $lower <= 122) {
                if (strlen($word) < $maxWord) {
                    $word .= chr($lower);
                }
                continue;
            }

            if ($word !== '') {
                $counts[$word] = ($counts[$word] ?? 0) + 1;
                $total++;
                $word = '';
            }
        }

        if ($word !== '') {
            $counts[$word] = ($counts[$word] ?? 0) + 1;
            $total++;
        }

        $entries = [];
        foreach ($counts as $entryWord => $count) {
            $entries[] = new Entry($entryWord, $count);
        }

        usort(
            $entries,
            static function (Entry $left, Entry $right): int {
                $countOrder = $right->count <=> $left->count;
                if ($countOrder !== 0) {
                    return $countOrder;
                }

                return strcmp($left->word, $right->word);
            },
        );

        return new Result($total, count($counts), array_slice($entries, 0, $top));
    }

    private static function normalizeMaxWord(int $value): int
    {
        if ($value === 0) {
            return self::ORACLE_DEFAULT_MAX_WORD;
        }

        return min(max($value, self::MIN_WORD), self::MAX_WORD_LIMIT);
    }
}
